lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -925,52 +1001,44 @@ public function andHaving($predicate, ...$predicates) : self
      */
     public function orHaving($predicate, ...$predicates) : self
     {
-        return $this->appendPredicates('having', CompositeExpression::TYPE_OR, $predicate, ...$predicates);
+        $this->having = $this->appendToPredicate($this->having, CompositeExpression::TYPE_OR, $predicate, ...$predicates);
+
+        $this->state = self::STATE_DIRTY;
+
+        return $this;
     }
 
     /**
-     * Sets one or more predicates combined by the AND logic as the given query clause.
-     * Replaces any previously specified predicates.
+     * Creates a CompositeExpression from one or more predicates combined by the AND logic.
      *
+     * @param string|CompositeExpression $predicate
      * @param string|CompositeExpression ...$predicates
-     *
-     * @return $this This QueryBuilder instance.
      */
-    private function setPredicates(string $clause, ...$predicates) : self
+    private function createPredicate($predicate, ...$predicates) : CompositeExpression
     {
-        if (count($predicates) > 1) {
-            $predicate = new CompositeExpression(
-                CompositeExpression::TYPE_AND,
-                $predicates
-            );
-        } else {
-            $predicate = array_shift($predicates);
+        if ($predicate instanceof CompositeExpression && ! $predicates) {
+            return $predicate;
         }
 
-        return $this->add($clause, $predicate);
+        return new CompositeExpression(CompositeExpression::TYPE_AND, array_merge([$predicate], $predicates));
     }
 
     /**
      * Appends the given predicates combined by the given type of logic to the given query clause.
      *
      * @param string|CompositeExpression ...$predicates
-     *
-     * @return $this This QueryBuilder instance.
      */
-    private function appendPredicates(string $clause, string $type, ...$predicates) : self
+    private function appendToPredicate(?CompositeExpression $predicate, string $type, ...$predicates) : CompositeExpression
     {
-        $predicate = $this->getQueryPart($clause);
-
-        if ($predicate instanceof CompositeExpression && $predicate->getType() === $type) {
-            $predicate->addMultiple($predicates);
-        } else {
-            $predicate = new CompositeExpression(
-                $type,
-                array_merge([$predicate], $predicates)
-            );
+        if ($predicate !== null) {
+            if ($predicate->getType() === $type) {
+                return $predicate->addMultiple($predicates);
```

Comment:
I'm a bit uncomfortable with this mutable `CompositeExpression`, I feel like it's a can of worms that can lead to some nasty bugs.

Example (pure speculation):

- I pass in a single `CompositeExpression` I created, to both `select()` and `having()`;
- This is already a `CompositeExpression`, single argument, so it's directly assigned to `$select` and `$having`;
- Now I `addSelect()`; the original `CompositeExpression` will be mutated, affecting both the value in `$select` and `$having`!

I haven't tested it, but I'm pretty sure there's a bug there, ready to be discovered. Again, let's not fix bugs & refactor in a single PR. If you wish, I can later open a PR to make `CompositeExpression` immutable, which will prevent any such issue.

Reply:
I had a very similar idea while working on this code. Let's put it aside and get back to it later.

Reply:
Filed https://github.com/doctrine/dbal/issues/3844.